File: Python/correlation/stats_corr_corrMatrix.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: Correlation
# ### VIDEO: Correlation matrix
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


# import libraries
import matplotlib.pyplot as plt
import numpy as np
import numpy.matlib as matlib


# In[ ]:


## simulate data

# simulation parameters
N = 1000  # time points
M =   20  # channels

# time vector (radian units)
t = np.linspace(0,6*np.pi,N)

# relationship across channels (imposing covariance)
chanrel = np.sin(np.linspace(0,2*np.pi,M))

# create the data
data = np.zeros((M,N))
for mi in range(M):
    data[mi:] = np.sin(t) * chanrel[mi]

data = data + np.random.randn(M,N)
    

# two ways of visualizing the multichannel data
for i in range(M):
    plt.plot(t,data[i,:]+i*4)
    
plt.yticks([])
plt.xlabel('Time (a.u.)')
plt.ylabel('Channel')
plt.show()

plt.imshow(data,aspect='auto',vmin=-2,vmax=2,extent=[t[0],t[-1],0,M])
plt.xlabel('Time (a.u.)')
plt.ylabel('Channel')
plt.show()


# In[ ]:


## now compute the covariance matrix

# note the size of the output!
dataCovMat = np.cov(data.T)

plt.imshow(dataCovMat,vmin=-.5,vmax=.5)
plt.title('Data covariance matrix')
plt.xlabel('??')
plt.ylabel('??')
plt.show()


# In[ ]:


## and now the correlation matrix

# note the size of the output!
dataCorrMat = np.corrcoef(data)

plt.imshow(dataCorrMat,vmin=-.5,vmax=.5)
plt.title('Data correlation matrix')
plt.xlabel('??')
plt.ylabel('??')
plt.show()

File: MATLAB/correlation/stats_corr_corrMatrix.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Correlation
%    VIDEO: Correlation matrix
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear; clc

%% simulate data

% simulation parameters
N = 1000;     % time points
M =   20;     % channels

% time vector (radian units)
t = linspace(0,6*pi,N);

% relationship across channels (imposing covariance)
chanrel = sin(linspace(0,2*pi,M))';

% create the data
data = bsxfun(@times,repmat( sin(t),M,1 ),chanrel);
data = data + randn(M,N);

% two ways of visualizing the multichannel data
figure(1), clf
plot(t,bsxfun(@plus,data,(1:M)'*4))
set(gca,'ytick',[],'xlim',t([1 end]))
xlabel('Time (a.u.)')
ylabel('Channel')

figure(2), clf
imagesc(t,[],data)
xlabel('Time (a.u.)')
ylabel('Channel')
set(gca,'clim',[-1 1]*2)

%% now compute the covariance matrix

% note the size of the output!
dataCovMat = cov(data');

figure(3), clf
imagesc(dataCovMat)
axis image
set(gca,'clim',[-1 1]*.5)
title('Data covariance matrix')
xlabel('??')
ylabel('??')

%% and now the correlation matrix

% note the size of the output!
dataCorrMat = corrcoef(data');

figure(4), clf
imagesc(dataCorrMat)
axis image
set(gca,'clim',[-1 1]*.5)
title('Data correlation matrix')
xlabel('??')
ylabel('??')

%% done.
